test/e2e/ingress.go

```diff
@@ -0,0 +1,311 @@
+		fmt.Sprintf("--regex=%v", regex), "-q", "--format=json").CombinedOutput()
+	if err != nil {
+		Failf("%v", err)
+	}
+	if err := json.Unmarshal([]byte(output), out); err != nil {
+		Failf("%v", err)
+	}
+}
+
+func checkLeakedResources() error {
+	msg := ""
+	// TODO: These are the main resources but we should really check them all.
+	beList := []compute.BackendService{}
+	gcloudUnmarshal("backend-services", "k8s-be-[0-9]+", &beList)
+	if len(beList) != 0 {
+		for _, b := range beList {
+			msg += fmt.Sprintf("%v\n", b.Name)
+		}
+		return fmt.Errorf("Found backend services:\n%v", msg)
+	}
+	fwList := []compute.ForwardingRule{}
+	gcloudUnmarshal("forwarding-rules", "k8s-fw-.*", &fwList)
+	if len(fwList) != 0 {
+		for _, f := range fwList {
+			msg += fmt.Sprintf("%v\n", f.Name)
+		}
+		return fmt.Errorf("Found forwarding rules:\n%v", msg)
+	}
+	return nil
+}
+
+var _ = Describe("GLBC", func() {
+	// These variables are initialized after framework's beforeEach.
+	var ns string
+	var client *client.Client
+	var responseTimes, creationTimes []time.Duration
+
+	framework := Framework{BaseName: "glbc"}
+
+	BeforeEach(func() {
+		// This test requires a GCE/GKE only cluster-addon
+		SkipUnlessProviderIs("gce", "gke")
+		framework.beforeEach()
+		client = framework.Client
+		ns = framework.Namespace.Name
+		Expect(waitForRCPodsRunning(client, "kube-system", "glbc")).NotTo(HaveOccurred())
+		Expect(checkLeakedResources()).NotTo(HaveOccurred())
+		responseTimes = []time.Duration{}
+		creationTimes = []time.Duration{}
+	})
+
+	AfterEach(func() {
+		framework.afterEach()
+		wait.Poll(lbPollInterval, lbPollTimeout, func() (bool, error) {
+			if err := checkLeakedResources(); err != nil {
+				Logf("Still waiting for glbc to cleanup: %v", err)
+				return false, nil
+			}
+			return true, nil
+		})
+
+		// In most cases slow loadbalancer creation/startup translates directly to
+		// GCE api sluggishness. However this might be because of something the
+		// controller is doing, eg: maxing out QPS by repeated polling.
+		sort.Sort(timeSlice(creationTimes))
+		perc50 := creationTimes[len(creationTimes)/2]
+		if perc50 > avgLBLatency {
+			Failf("Average creation time is too high: %+v", creationTimes)
+		}
+		By(fmt.Sprintf("Loadbalancer creation latencies %+v", creationTimes))
+
+		sort.Sort(timeSlice(responseTimes))
+		perc50 = responseTimes[len(responseTimes)/2]
+		if perc50 > avgLBLatency {
+			Failf("Average startup time is too high: %+v", responseTimes)
+		}
+		By(fmt.Sprintf("Loadbalancer startup latencies %+v", responseTimes))
+	})
+
+	It("should create GCE L7 loadbalancers through the GLBC cluster addon", func() {
+		// Create numApps apps, exposed via numIng Ingress each with 2 paths.
+		// Eg with numApp=10, numIng=5:
+		// apps: {foo-app-(0-10)}
+		// ingress: {foo-app-(0, 2, 4, 6, 8)}
+		// paths:
+		//  ingress foo-app-0:
+		//	  default1.bar.com
+		//	  foo0.bar.com: /foo0
+		//	  foo1.bar.com: /foo1
+		if numApps < numIng {
```

Comment:
I think i'd have to encode it as Expect(numApps<numIng).NotToEqual(false) or something, which is weird